### ISSUE TITLE
SpinButton does not check for defaultValue on render method after setting state

### ISSUE THREAD

**Issue by katiearriagam** [microsoft/fluentui#13481]
### Environment Information

- **Package version(s)**: 7.117.0
- **Browser and OS versions**: Not relevant to this bug

### Please provide a reproduction of the bug in a codepen: https://codepen.io/katiearriagam/pen/MWKYOjy

Repro:
1. Notice the first SpinButton sets the value to `2` on render.
2. Click on the PrimaryButton - the handler `_setDefaultValue` sets the `defaultValue` to `""` and triggers a re-render because it does so through `setState`.
3. Notice also that on rendering the SpinButton, we set the `value` to `undefined` if we receive a defined `defaultValue`. 

#### Actual behavior:

- The SpinButton does not update the value after clicking on the button. It still reads `2` as the value of the SpinButton.
- **Root cause:** when we set the defaultValue, the `render()` method in SpinButton does not check for defaultValue as it does in the constructor.

#### Expected behavior:

- In the CodePen, the SpinButton should use the defaultValue `""` instead of the previous value after setting the state (after button click).
- The `render()` method should check for `defaultValue` in the same way that this check exists in the constructor.
```tsx
  // Don't use || here because it won't handle empty strings properly
    let { value = props.defaultValue } = props;
    if (value === undefined) {
      value = typeof props.min === 'number' ? String(props.min) : '0';
    }
   this._lastValidValue = value;
```

### Priorities and help requested:

Are you willing to submit a PR to fix? No

Requested priority: Blocking

Products/sites affected: Office Online UI

**Comment by katiearriagam** [comment]
FYI: @jspurlin

**Comment by katiearriagam** [comment]
Also, I am aware that the documentation states: 

```tsx
  /**
   * Initial value of the control. Updates to this prop will not be respected.
   *
   * Use this if you intend for the SpinButton to be an uncontrolled component which maintains its own value.
   * Mutually exclusive with `value`.
   * @defaultvalue 0
   */
  defaultValue?: string;
```

but I'm not sure why this restriction exists. In our Office Online UI scenario, the default value changes based on the selection for another component.

**Comment by jdhuntington** [comment]
Talked offline about next steps. Unfortunately cannot make any meaningful changes in v7 as it would be a breaking change. Keeping this issue open for discussion as v8 is undergoing planning and design.

**Comment by ecraig12345** [comment]
This is already behaving according to the expected controlled/uncontrolled component pattern (at least in this case), which mirrors React's behavior with built-in `<input>` elements.

**Uncontrolled** components internally track what the user entered. You can optionally provide a `defaultValue` prop to start from, but updates to `defaultValue` shouldn't be respected. (A bunch of our components handle this incorrectly at the moment.)

**Controlled** components get their current value from `props.value` and notify of changes via `onChange` or similar. This means you have full control of the value shown but are also responsible for any state management and value updates. These components **should not** allow updates apart from `props.value` changes (though this is often implemented incorrectly, including by SpinButton currently).

There's also a workaround to use `defaultValue` but occasionally allow updates by changing `key` to re-create the component.

[More details on all of that and links to example code here.](https://gist.github.com/ecraig12345/1ea4566ce171b844a0af5a3366d3473f)

One note specific to SpinButton: implementing the controlled pattern is currently pretty awkward since SpinButton doesn't have an actual `onChange` handler (so you have to listen to and re-implement `onIncrement`, `onDecrement`, and `onValidate`), but this will be fixed in version 8--see #5326.
